### PR TITLE
Disallow remote deploy for typed actors

### LINKED ISSUE #24352: Disallow remote deployment in typed
Since we don't want to support it for typed actors we should make sure that it's not technically possible. I think it currently is, with configuration.

### PR DESCRIPTION
**Pull request by johanandren**
Fixes #24352

### CODE REVIEW

**Comment by johanandren** [comment]
Haven't figured out how I could prevent it for a typed child to an untyped actor, where remote deploy is programatically added to the Props returned by PropsAdapter

File: akka-actor/src/main/scala/akka/actor/Deployer.scala [hunk 1]
```diff
@@ -109,6 +110,16 @@ case object LocalScope extends LocalScope {
   def withFallback(other: Scope): Scope = this
 }
 
+/**
+ * INTERNAL API
+ *
+ * Used for disallowing remote deployment for adapted typed actors
+ */
+@InternalApi
+private[akka] case object DisallowRemoteScope extends LocalScope {
```
Comment (patriknw) [inline akka-actor/src/main/scala/akka/actor/Deployer.scala]: this isn't used (yet)?

File: akka-cluster-typed/src/test/scala/akka/cluster/typed/RemoteDeployNotAllowedSpec.scala [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+      }
+      cluster.jmx.enabled = false
+    }
+    """)
+
+  def configWithRemoteDeployment(otherSystemPort: Int) = ConfigFactory.parseString(
+    s"""
+      akka.actor.deployment {
+        "/*" {
+          remote = "akka.tcp://sampleActorSystem@127.0.0.1:$otherSystemPort"
```
Comment (patriknw) [inline akka-cluster-typed/src/test/scala/akka/cluster/typed/RemoteDeployNotAllowedSpec.scala]: artery is enabled above, but this is using tcp?

File: akka-cluster-typed/src/test/scala/akka/cluster/typed/RemoteDeployNotAllowedSpec.scala [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+              ctx.spawnAnonymous(Behaviors.deferred[AnyRef] { ctx ⇒ Behaviors.empty })
+            } catch {
+              case ex: Exception ⇒ probe.ref ! ex
+            }
+            Behaviors.same
+        }
+
+      }
+
+      val system2 = ActorSystem(guardianBehavior, system.name,
```
Comment (patriknw) [inline akka-cluster-typed/src/test/scala/akka/cluster/typed/RemoteDeployNotAllowedSpec.scala]: remember to terminate system2

**Review by patriknw — COMMENTED** [review]
looks like a simple (good) approach

File: akka-actor/src/main/scala/akka/actor/Deployer.scala (hunk 1 above)
Comment (johanandren) [reply akka-actor/src/main/scala/akka/actor/Deployer.scala]: Argh, tired Friday PR, was from my previous attempt to solve it.

File: akka-cluster-typed/src/test/scala/akka/cluster/typed/RemoteDeployNotAllowedSpec.scala (hunk 3 above)
Comment (ktoso) [reply akka-cluster-typed/src/test/scala/akka/cluster/typed/RemoteDeployNotAllowedSpec.scala]: Seems to be addressed

**Review by ktoso — APPROVED** [review]
Looks good. LGTM

File: akka-actor-typed/src/main/scala/akka/actor/typed/internal/adapter/ActorContextAdapter.scala [hunk 4]
```diff
@@ -102,8 +103,13 @@ import akka.actor.typed.Behavior.UntypedBehavior
...
         ActorRefAdapter(ctx.actorOf(b.untypedProps))
       case _ ⇒
-        Behavior.validateAsInitial(behavior)
-        ActorRefAdapter(ctx.actorOf(PropsAdapter(() ⇒ behavior, props)))
+        try {
+          Behavior.validateAsInitial(behavior)
+          ActorRefAdapter(ctx.actorOf(PropsAdapter(() ⇒ behavior, props)))
+        } catch {
+          case ex: ConfigurationException if ex.getMessage.startsWith("configuration requested remote deployment") ⇒
+            throw new ConfigurationException("Remote deployment not allowed for typed actors", ex)
```
Comment (patriknw) [inline akka-actor-typed/src/main/scala/akka/actor/typed/internal/adapter/ActorContextAdapter.scala]: not so nice to rely on the exception messages, but there is a test that will fail if we change

**Review by patriknw — APPROVED** [review]
LGTM

**Pull request merged**